lib/indexer/addrindexer.js

```diff
@@ -0,0 +1,279 @@
+ */
+
+class AddrIndexer extends Indexer {
+  /**
+   * Create a indexer
+   * @constructor
+   * @param {Object} options
+   */
+
+  constructor(options) {
+    super('addr', options);
+
+    this.db = bdb.create(this.options);
+    this.maxTxs = options.maxTxs || 100;
+  }
+
+  /**
+   * Index transactions by address.
+   * @private
+   * @param {BlockMeta} meta
+   * @param {Block} block
+   * @param {CoinView} view
+   */
+
+  async indexBlock(meta, block, view) {
+    const height = meta.height;
+
+    for (let i = 0; i < block.txs.length; i++) {
+      const tx = block.txs[i];
+      const hash = tx.hash();
+      const count = new Count(height, i);
+
+      let hasAddress = false;
+
+      for (const addr of tx.getAddresses(view)) {
+        const prefix = addr.getPrefix();
+
+        if (prefix < 0)
+          continue;
+
+        const addrHash = addr.getHash();
+
+        this.put(layout.A.encode(prefix, addrHash, height, i), null);
+
+        hasAddress = true;
+      }
+
+      if (hasAddress) {
+        this.put(layout.C.encode(height, i), hash);
+        this.put(layout.c.encode(hash), count.toRaw());
+      }
+    }
+  }
+
+  /**
+   * Remove addresses from index.
+   * @private
+   * @param {BlockMeta} meta
+   * @param {Block} block
+   * @param {CoinView} view
+   */
+
+  async unindexBlock(meta, block, view) {
+    const height = meta.height;
+
+    for (let i = 0; i < block.txs.length; i++) {
+      const tx = block.txs[i];
+      const hash = tx.hash();
+
+      let hasAddress = false;
+
+      for (const addr of tx.getAddresses(view)) {
+        const prefix = addr.getPrefix();
+
+        if (prefix < 0)
+          continue;
+
+        const addrHash = addr.getHash();
+
+        this.del(layout.A.encode(prefix, addrHash, height, i));
+
+        hasAddress = true;
+      }
+
+      if (hasAddress) {
+        this.del(layout.C.encode(height, i));
+        this.del(layout.c.encode(hash));
+      }
+    }
+  }
+
+  /**
+   * Get transaction hashes to an address in ascending or descending
+   * order. If the `after` argument is supplied, results will be given
+   * _after_ that transaction hash. The default order is ascending from
+   * oldest to latest.
+   * @param {Address} addr
+   * @param {Object} options
+   * @param {Buffer} options.after - A transaction hash
+   * @param {Number} options.limit
+   * @param {Boolean} options.reverse
+   * @returns {Promise} - Returns {@link Hash}[].
+   */
+
+  async getHashesByAddress(addr, options = {}) {
+    const txs = [];
+
+    const {after, reverse} = options;
+    let {limit} = options;
+
+    if (!limit)
+      limit = this.maxTxs;
+
+    if (limit > this.maxTxs)
+      throw new Error('Limit above max of ${this.maxTxs}.');
+
+    const hash = Address.getHash(addr);
+    const prefix = addr.getPrefix();
+
+    const opts = {
+      limit,
+      reverse,
+      parse: (key) => {
+        const [,, height, index] = layout.A.decode(key);
+        txs.push([height, index]);
```

Comment:
instead of mutating txs here, returning them and having `const txs = await this.db.keys...` will be more readable

Reply:
I think that makes sense.